File: .agents/validators/qala_ksa_pii.py
# SPDX-License-Identifier: MIT
# Licensed under MIT
"""Qal'a (قلعة) — Q5 Sovereign KSA-PII Detector (Python side).

Detects Saudi-specific identifiers in text payloads. Distinct from
generic secret detection in ``sovereignCyberRadar.SECRET_PATTERNS`` and
adapter ``SENSITIVE_PATTERNS``.

Raw match values are NEVER returned. Each hit is reported as
``KsaPiiHit(category, masked_value, span)`` where ``masked_value``
preserves only the first/last two characters with the middle replaced
by ``…``. Callers may pass results to the Q7 audit sink without
re-redaction.

No network calls. No persistent state. No background workers.

Mirror of ``src/security/qalaKsaPii.ts``. See
``docs/decisions/ADR-0003-qala-security-architecture.md`` §Q5.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Final, Literal

KsaPiiCategory = Literal[
    "KSA_NATIONAL_ID",
    "KSA_IQAMA",
    "KSA_IBAN",
    "KSA_MOBILE",
    "KSA_COMMERCIAL_REGISTRATION",
    "KSA_ID_AMBIGUOUS_10DIGIT",
]


@dataclass(frozen=True)
class KsaPiiHit:
    category: KsaPiiCategory
    masked_value: str
    span: tuple[int, int]


# ── Patterns ────────────────────────────────────────────────────────────────
#
# Word-boundary discipline matters: 10-digit shapes (National ID, Iqama, CR)
# overlap heavily, so we anchor with (?<!\d) and (?!\d) to avoid matching
# substrings of longer digit runs. Leading-digit discrimination follows:
#   National ID  : leading 1
#   Iqama        : leading 2
#   CR           : leading 7 (common KSA CR prefix); ambiguous 10-digit
#                  shapes outside {1,2,7} are reported as KSA_ID_AMBIGUOUS_10DIGIT.

_IBAN_RE: Final[re.Pattern[str]] = re.compile(r"\bSA\d{22}\b")
_NATIONAL_ID_RE: Final[re.Pattern[str]] = re.compile(r"(?<!\d)1\d{9}(?!\d)")
_IQAMA_RE: Final[re.Pattern[str]] = re.compile(r"(?<!\d)2\d{9}(?!\d)")
_CR_RE: Final[re.Pattern[str]] = re.compile(r"(?<!\d)7\d{9}(?!\d)")
_AMBIGUOUS_10DIGIT_RE: Final[re.Pattern[str]] = re.compile(
    r"(?<!\d)[3-689]\d{9}(?!\d)"
)
# Saudi mobile: +9665XXXXXXXX | 9665XXXXXXXX | 05XXXXXXXX (10 digits with leading 0)
_MOBILE_RE: Final[re.Pattern[str]] = re.compile(
    r"(?<!\d)(?:\+?966\s?5\d{8}|05\d{8})(?!\d)"
)


def _mask(value: str) -> str:
    """Preserve first/last two chars; replace middle with ``…``.

    Examples:
        "1234567890"     -> "12…90"
        "SA0000000000…"  -> "SA…<last2>"
        "05XXXXXXXX"     -> "05…<last2>"
    For inputs shorter than 4 chars, returns ``"…"`` to avoid leaking
    any character at all.
    """
    if len(value) < 4:
        return "…"
    return f"{value[:2]}…{value[-2:]}"


def _scan(
    pattern: re.Pattern[str],
    text: str,
    category: KsaPiiCategory,
) -> list[KsaPiiHit]:
    return [
        KsaPiiHit(category=category, masked_value=_mask(m.group(0)), span=m.span())
        for m in pattern.finditer(text)
    ]


def detect_ksa_pii(text: str) -> tuple[KsaPiiHit, ...]:
    """Return all detected KSA-PII hits, deduplicated by span.

    Categories are anti-collision: IBAN matches do not double-count as
    10-digit IDs, and 10-digit IDs are partitioned by leading digit
    into National ID / Iqama / CR / Ambiguous.
    """
    if not isinstance(text, str) or len(text) == 0:
        return ()

    hits: list[KsaPiiHit] = []
    consumed: set[tuple[int, int]] = set()

    # IBAN first — it's the most specific shape and would otherwise
    # be mis-classified as a number inside an alphanumeric blob.
    for hit in _scan(_IBAN_RE, text, "KSA_IBAN"):
        if hit.span not in consumed:
            hits.append(hit)
            consumed.add(hit.span)

    # Mobile next — its anchoring includes country-code shapes that
    # do not overlap with the 10-digit ID anchors.
    for hit in _scan(_MOBILE_RE, text, "KSA_MOBILE"):
        if hit.span not in consumed:
            hits.append(hit)
            consumed.add(hit.span)

    # 10-digit IDs partitioned by leading digit.
    for pattern, category in (
        (_NATIONAL_ID_RE, "KSA_NATIONAL_ID"),
        (_IQAMA_RE, "KSA_IQAMA"),
        (_CR_RE, "KSA_COMMERCIAL_REGISTRATION"),
        (_AMBIGUOUS_10DIGIT_RE, "KSA_ID_AMBIGUOUS_10DIGIT"),
    ):
        for hit in _scan(pattern, text, category):  # type: ignore[arg-type]
            if hit.span not in consumed:
                hits.append(hit)
                consumed.add(hit.span)

    return tuple(sorted(hits, key=lambda h: h.span))


def has_ksa_pii(text: str) -> bool:
    return len(detect_ksa_pii(text)) > 0


def redact_ksa_pii(text: str) -> str:
    """Return ``text`` with every KSA-PII span replaced by its mask.

    Raw values never appear in the output.
    """
    hits = detect_ksa_pii(text)
    if not hits:
        return text
    pieces: list[str] = []
    cursor = 0
    for hit in hits:
        start, end = hit.span
        pieces.append(text[cursor:start])
        pieces.append(f"[{hit.category}:{hit.masked_value}]")
        cursor = end
    pieces.append(text[cursor:])
    return "".join(pieces)


__all__ = [
    "KsaPiiCategory",
    "KsaPiiHit",
    "detect_ksa_pii",
    "has_ksa_pii",
    "redact_ksa_pii",
]

File: src/security/qalaKsaPii.js
// Qal'a (قلعة) — Q5 Sovereign KSA-PII Detector (Node side).
//
// Detects Saudi-specific identifiers in text payloads. Distinct from
// generic secret detection in sovereignCyberRadar.SECRET_PATTERNS and
// adapter SENSITIVE_PATTERNS.
//
// Raw match values are NEVER returned. Each hit is reported as
// { category, maskedValue, span } where maskedValue preserves only the
// first/last two characters with the middle replaced by "…". Callers
// may pass results to the Q7 audit sink without re-redaction.
//
// No network calls. No persistent state. No background workers.
//
// See docs/decisions/ADR-0003-qala-security-architecture.md §Q5.
// Word-boundary discipline: 10-digit shapes (National ID, Iqama, CR)
// overlap heavily, so we anchor with (?<!\d) and (?!\d) to avoid
// matching substrings of longer digit runs.
const IBAN_RE = /\bSA\d{22}\b/g;
const NATIONAL_ID_RE = /(?<!\d)1\d{9}(?!\d)/g;
const IQAMA_RE = /(?<!\d)2\d{9}(?!\d)/g;
const CR_RE = /(?<!\d)7\d{9}(?!\d)/g;
const AMBIGUOUS_10DIGIT_RE = /(?<!\d)[3-689]\d{9}(?!\d)/g;
const MOBILE_RE = /(?<!\d)(?:\+?966\s?5\d{8}|05\d{8})(?!\d)/g;
function maskValue(value) {
    if (value.length < 4)
        return "…";
    return `${value.slice(0, 2)}…${value.slice(-2)}`;
}
function scan(source, pattern, category) {
    const hits = [];
    // Always reset lastIndex before iterating to keep the function pure.
    pattern.lastIndex = 0;
    let match = pattern.exec(source);
    while (match !== null) {
        const start = match.index;
        const end = start + match[0].length;
        hits.push({
            category,
            maskedValue: maskValue(match[0]),
            span: [start, end],
        });
        match = pattern.exec(source);
    }
    return hits;
}
export function detectKsaPii(text) {
    if (typeof text !== "string" || text.length === 0)
        return [];
    const consumed = new Set();
    const hits = [];
    function consider(candidates) {
        for (const hit of candidates) {
            const key = `${hit.span[0]}:${hit.span[1]}`;
            if (!consumed.has(key)) {
                consumed.add(key);
                hits.push(hit);
            }
        }
    }
    // Most specific first: IBAN, then mobile (with country-code shape),
    // then 10-digit IDs partitioned by leading digit.
    consider(scan(text, IBAN_RE, "KSA_IBAN"));
    consider(scan(text, MOBILE_RE, "KSA_MOBILE"));
    consider(scan(text, NATIONAL_ID_RE, "KSA_NATIONAL_ID"));
    consider(scan(text, IQAMA_RE, "KSA_IQAMA"));
    consider(scan(text, CR_RE, "KSA_COMMERCIAL_REGISTRATION"));
    consider(scan(text, AMBIGUOUS_10DIGIT_RE, "KSA_ID_AMBIGUOUS_10DIGIT"));
    hits.sort((a, b) => a.span[0] - b.span[0]);
    return hits;
}
export function hasKsaPii(text) {
    return detectKsaPii(text).length > 0;
}
export function redactKsaPii(text) {
    const hits = detectKsaPii(text);
    if (hits.length === 0)
        return text;
    const pieces = [];
    let cursor = 0;
    for (const hit of hits) {
        const [start, end] = hit.span;
        pieces.push(text.slice(cursor, start));
        pieces.push(`[${hit.category}:${hit.maskedValue}]`);
        cursor = end;
    }
    pieces.push(text.slice(cursor));
    return pieces.join("");
}
